app/Models/NetworkCampaign.php

```diff
@@ -0,0 +1,167 @@
+     *
+     * @var array
+     */
+    protected $fillable = [
+      'uuid',
+      'source_created_at', 'source_updated_at',
+      'source_host', 'adshares_address',
+      'landing_url', 'max_cpm', 'max_cpc', 'budget_per_hour', 'time_start', 'time_end',
+    ];
+
+    /**
+     * The attributes that should be hidden for arrays.
+     *
+     * @var array
+     */
+    protected $hidden = [
+      'id'
+    ];
+
+    /**
+    * The attributes that use some Models\Traits with mutator settings automation
+    *
+    * @var array
+    */
+    protected $traitAutomate = [
+      'uuid' => 'BinHex',
+    ];
+
+    public function banners()
+    {
+        return $this->hasMany('Adshares\Adserver\Models\NetworkBanner');
+    }
+
+    public function campaignExcludes()
+    {
+        return $this->hasMany('Adshares\Adserver\Models\NetworkCampaignExclude');
+    }
+
+    public function campaignRequires()
+    {
+        return $this->hasMany('Adshares\Adserver\Models\NetworkCampaignRequire');
+    }
+
+    protected static function convertTimestampsToSourceTimestamps($array)
+    {
+        $array['source_created_at'] = $array['created_at'];
+        $array['source_updated_at'] = $array['updated_at'];
+        unset($array['created_at']);
+        unset($array['updated_at']);
+        return $array;
+    }
+
+    protected static function jsonDataMakeUUIDKeys(array &$data)
+    {
+        foreach ($data['banners'] as $i => $d) {
+            $data['banners'][$d['uuid']] = self::convertTimestampsToSourceTimestamps($d);
+            unset($data['banners'][$i]);
+        }
+        foreach ($data['campaign_excludes'] as $i => $d) {
+            $data['campaign_excludes'][$d['uuid']] = self::convertTimestampsToSourceTimestamps($d);
+            unset($data['campaign_excludes'][$i]);
+        }
+        foreach ($data['campaign_requires'] as $i => $d) {
+            $data['campaign_requires'][$d['uuid']] = self::convertTimestampsToSourceTimestamps($d);
+            unset($data['campaign_requires'][$i]);
+        }
+    }
+
+    public static function fromJsonData(array $data)
+    {
+        DB::beginTransaction();
+
+        // print_r($data);
+
+        $campaign = self::with('Banners', 'CampaignExcludes', 'CampaignRequires')->where('uuid', hex2bin($data['uuid']))->lockForUpdate()->first();
+
+        if (empty($campaign)) {
+            $campaign = self::fromJsonDataNew($data);
+        } else {
+            $campaign = self::fromJsonDataUpdate($campaign, $data);
+        }
+
+        // print_r($campaign->toArray());
+
+        DB::commit();
+        // DB::rollBack();
+    }
+
+    public static function fromJsonDataNew(array $data)
+    {
+        $data = self::convertTimestampsToSourceTimestamps($data);
+        $campaign = self::create($data);
+
+        foreach ($data['banners'] as $d) {
+            $d = self::convertTimestampsToSourceTimestamps($d);
+            $d['network_campaign_id'] = $campaign->id;
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "network_campaign_id" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)